improver/regrid/landsea2.py

```diff
@@ -0,0 +1,208 @@
+        """
+        Initialise class
+        """
+        self.regrid_mode = regrid_mode
+        self.vicinity = vicinity_radius
+
+    def process(self, cube_in, cube_in_mask, cube_out_mask):
+        """
+        Regridding considering land_sea mask. please note cube_in must use
+        lats/lons rectlinear system(GeogCS). cube_in_mask and cube_in could be
+        different  resolution. cube_our could be either in lats/lons rectlinear
+        system or LambertAzimuthalEqualArea system.
+
+        Args:
+            cube_in (iris.cube.Cube):
+                Cube of data to be regridded
+            cube_in_mask (iris.cube.Cube):
+                Cube of land_binary_mask data ((land:1, sea:0). used to determine
+                where the input model data is representing land and sea points.
+            cube_out_mask (iris.cube.Cube):
+                Cube of land_binary_mask data on target grid (land:1, sea:0).
+
+        Returns:
+            iris.cube.Cube:
+                Regridded result cube
+        """
+
+        # Gather output latitude/longitudes from output template cube
+        if (
+            cube_out_mask.coord(axis="x").standard_name == "projection_x_coordinate"
+            and cube_out_mask.coord(axis="y").standard_name == "projection_y_coordinate"
+        ):
+            out_latlons = convert_from_projection_to_latlons(cube_out_mask, cube_in)
+        else:
+            out_latlons = latlon_from_cube(cube_out_mask)
+
+        # Subset the input cube so that extra spatial area beyond the output is removed
+        # This is a performance optimisation to reduce the size of the dataset being processed
+        lat_max, lon_max = out_latlons.max(axis=0)
+        lat_min, lon_min = out_latlons.min(axis=0)
+        if WITH_MASK in self.regrid_mode:
+            cube_in, cube_in_mask = slice_mask_cube_by_domain(
+                cube_in, cube_in_mask, (lat_max, lon_max, lat_min, lon_min)
+            )
+        else:  # not WITH_MASK
+            cube_in = slice_cube_by_domain(
+                cube_in, (lat_max, lon_max, lat_min, lon_min)
+            )
+
+        # Gather input latitude/longitudes from input cube
+        in_latlons = latlon_from_cube(cube_in)
+        # Number of grid points in X dimension is used to work out length of flattened array
+        # stripes for finding surrounding points for bilinear interpolation
+        in_lons_dim = cube_in.coord(axis="x").shape[0]  # longitude
+
+        # Locate nearby input points for output points
+        indexes = basic_indexes(out_latlons, in_latlons, in_lons_dim)
+        # Initialise distances and weights to zero. Weights are only used for the bilinear case
+        distances = np.zeros((out_latlons.shape[0], 4))
+        weights = np.zeros((out_latlons.shape[0], 4), dtype=np.float32)
+        # Fill in distances and weights
+        if NEAREST in self.regrid_mode:
+            for i in range(4):
```

Comment:
Perhaps here the "4" could be a global variable to avoid ["magic numbers"](https://en.wikipedia.org/wiki/Magic_number_(programming)).

Reply:
added NUM_NEIGHBOURS = 4